src/jsMain/kotlin/io/beatmaps/modreview/entry.kt

```diff
@@ -30,142 +30,166 @@ import react.dom.p
 import react.dom.td
 import react.dom.textarea
 import react.dom.tr
-import react.fc
-import react.useContext
-import react.useState
+import react.setState
+import kotlin.js.Promise
 
-external interface ModReviewEntryProps : Props {
-    var entry: ReviewDetail?
+external interface ModReviewEntryProps<T : CommentDetail> : Props {
+    var entry: T?
     var setUser: (String) -> Unit
+    var onDelete: (String) -> Promise<*>
+    var onSave: (ReviewSentiment?, String) -> Promise<AxiosResponse<ActionResponse>>?
 }
 
-val modReviewEntryRenderer = fc<ModReviewEntryProps> { props ->
-    val reasonRef = createRef<HTMLTextAreaElement>()
-    val (hidden, setHidden) = useState(false)
-    val (editing, setEditing) = useState(false)
-    val (sentiment, setSentiment) = useState(null as ReviewSentiment?)
-    val (newSentiment, setNewSentiment) = useState(null as ReviewSentiment?)
-    val (text, setText) = useState(null as String?)
+external interface ModReviewEntryState : State {
+    var hidden: Boolean
+    var editing: Boolean
+    var sentiment: ReviewSentiment?
+    var newSentiment: ReviewSentiment?
+    var text: String?
+}
 
-    val modal = useContext(modalContext)
+class ModReviewEntry<T : CommentDetail> : RComponent<ModReviewEntryProps<T>, ModReviewEntryState>() {
```

Comment:
If this needs to be a class component that's fine but after going through and trying to make all the components functional I'm wondering if it needs to be.

Did some messing around locally and the following pattern might work? If not I guess we might be stuck with class components.

```kt
fun <T: CommentDetail> makeModReviewEntry() =
    fc<ModReviewEntryProps<T>> { props ->
        ... code goes here ...
    }
fun <T : CommentDetail> RBuilder.modReviewEntry(handler: ModReviewEntryProps<T>.() -> Unit) =
    makeModReviewEntry<T>().invoke {
        attrs(handler)
    }
```

Reply:
Alternatively does anything actually need this to be generic at all?
The only places you're using entry you check the type anyway with `entry is ReviewDetail`

Removing all the generics also leads to no errors locally.